pkg/streamingpromql/operators/binary_operation.go

```diff
@@ -418,6 +446,42 @@ func (b *BinaryOperation) mergeOneSide(data []types.InstantVectorSeriesData, sou
 		return types.InstantVectorSeriesData{}, nil
 	}
 
+	// Merge floats and histograms individually.
+	// After which we check if there are any duplicate points in either the floats or histograms.
+	filteredData := filteredData{}
+	filteredData.filter(data)
+
+	floats, err := b.mergeOneSideFloats(filteredData, sourceSeriesIndices, sourceSeriesMetadata, side)
+	if err != nil {
+		return types.InstantVectorSeriesData{}, err
+	}
+	histograms, err := b.mergeOneSideHistograms(filteredData, sourceSeriesIndices, sourceSeriesMetadata, side)
+	if err != nil {
+		return types.InstantVectorSeriesData{}, err
+	}
+
+	// Check for any conflicts between floats and histograms
+	i, j := 0, 0
+	for i < len(floats) && j < len(histograms) {
+		if floats[i].T == histograms[j].T {
+			// Conflict found
+			return types.InstantVectorSeriesData{}, fmt.Errorf("found both float and histogram samples for the match group FIXME on the %s side of the operation at timestamp %s", side, timestamp.Time(floats[i].T))
+		}
+		if floats[i].T < histograms[j].T {
+			i++
+		} else {
+			j++
+		}
+	}
+
+	return types.InstantVectorSeriesData{Floats: floats, Histograms: histograms}, nil
+}
+
+func (b *BinaryOperation) mergeOneSideFloats(filteredData filteredData, sourceSeriesIndices []int, sourceSeriesMetadata []types.SeriesMetadata, side string) ([]promql.FPoint, error) {
+	data := filteredData.dataWithFloats
```

Comment:
Rather than doing this filtering in `filteredData`, what if we changed `mergeOneSideFloats` to run roughly like this:

1. if there are no input series at all, we're done
2. if there's one input series, we can just return that as-is, even if it's empty
3. sort all input series by timestamp of first float sample, sorting all series with no floats last
4. if, after sorting, the first series has no floats, we're done, as that means there are no floats in any series
5. check for overlaps between series, stopping once we reach a series with no samples
6. if we have no overlaps, we can just concatenate the slices from `data` together, ignoring the series with no float samples (we could do this by reslicing `data` to only the series with float samples)
7. if we have overlaps, we go through the process of merging them together, ignoring the series with no float samples (same as for 6, we could do this by reslicing `data` to only the series with float samples)

In the common case (one series on each side, either entirely with floats or entirely with histograms), this should be fast, and for the next most common case (multiple series without overlaps), this should still be fast, and we avoid the housekeeping of `filteredData`.